Fix/replace refs for relationships

barnabasJ opened the pull request:
# Contributor checklist

- [ ] Bug fixes include regression tests
- [ ] Chores
- [ ] Documentation changes
- [ ] Features include unit/acceptance tests
- [ ] Refactoring
- [ ] Update dependencies

barnabasJ commented on lib/ash/policy/authorizer/authorizer.ex:
```diff
@@ -712,6 +712,25 @@ defmodule Ash.Policy.Authorizer do
+        Ash.Resource.Info.primary_action!(resource, :read)
+
+    {expr, acc} =
+      expression_for_ref(resource, name, action, ref, acc)
```
We might also need to update the stack here. But doing it caused an error for me. 

I hit this case when coming from the `%{__function__?: true, ... }` above. Returning the original stack there might be the correct way to do things. What do you think?

zachdaniel replied:
🤔 I'm pretty sure the way you have it here is the right way :)

zachdaniel commented:
This looks good, *but* I don't understand exactly how we got to the scenario, a failing test would be great. If you have time. If not, we can merge.

zachdaniel commented:
I think it should be fine because `expression_for_ref` doesn't use the `stack` IIRC.

barnabasJ commented:
I tried adding a test. I did it in AshPostgres because I needed a filter function e.g `ilke` to be available. But I'm still having a hard time triggering the problem. Right now it does not call do_replace_refs because the `input?` flag on the Ref is not true. How do I need to set the filter to make it so?

barnabasJ commented:
> I tried adding a test. I did it in AshPostgres because I needed a filter function e.g `ilke` to be available. But I'm still having a hard time triggering the problem. Right now it does not call do_replace_refs because the `input?` flag on the Ref is not true. How do I need to set the filter to make it so?

I was able to figure it out by looking at the ash_graphql resolver, still having issues replicating the error though

barnabasJ commented:
I added this test in ash_postgres https://github.com/ash-project/ash_postgres/pull/206

The behaviour is a bit strange over all, but this way, you should be able to get a better picture of it.

barnabasJ commented:
> I added this test in ash_postgres [ash-project/ash_postgres#206](https://github.com/ash-project/ash_postgres/pull/206)
> 
> The behaviour is a bit strange over all, but this way, you should be able to get a better picture of it.

I added another case that is two relationships deep. This still fails even with my current change.

zachdaniel commented on lib/ash/policy/authorizer/authorizer.ex:
```diff
@@ -712,6 +712,25 @@ defmodule Ash.Policy.Authorizer do
+  defp do_replace_ref(
+         %{
+           attribute: %struct{name: name},
+           relationship_path: [relationship_path],
```
```suggestion
           relationship_path: relationship_path,
```

zachdaniel commented:
🚀 Thank you for your contribution! 🚀

This looked like some hard work to narrow this done. Fixed one small issue in your PR but otherwise this is good to go!
